experimental/segmenter/src/language.rs

```diff
@@ -19,3 +24,135 @@ pub fn get_language(codepoint: u32) -> Language {
         _ => Language::Unknown,
     }
 }
+
+/// This struct is an iterator that returns the string per language from the
+/// given string.
+///
+/// Actually supported LSTM model is Thai and Burmese only. If using other
+/// code point, it causes panic.
+pub struct LanguageIterator<'s> {
+    input: Chars<'s>,
+    current_ch: Option<char>,
+}
+
+impl<'s> LanguageIterator<'s> {
+    #[allow(dead_code)]
+    pub fn new(input: &'s str) -> Self {
+        let mut input = input.chars();
+        let current_ch = input.next();
+        Self { input, current_ch }
+    }
+}
+
+impl<'s> Iterator for LanguageIterator<'s> {
+    type Item = String;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut s = "".to_string();
+
+        let lang = get_language(self.current_ch? as u32);
+        s.push(self.current_ch.unwrap());
+        for c in self.input.by_ref() {
+            self.current_ch = Some(c);
+            let new_lang = get_language(c as u32);
+            if lang != new_lang {
+                return Some(s);
+            }
+            s.push(c);
+        }
+        self.current_ch = None;
+        Some(s)
+    }
```

Comment:
EOF or different language codepoint  is detected.

Reply:
I'm still confused. different language will result in `Some(s)`, but when does the iterator end and returns `None`?

If `self.input.by_ref()` doesn't return, then `s` is `"".to_string()` + `self.current_ch.unwrap()` (which may panic?) but it doesn't seem like there is a condition where the iterator returns `None`.

What am I missing?

Reply:
I see, ok, this code is quite cryptic, and we should revisit the logic here. I think it can be written in a more readable way, and it also seems like it should be able to not allocate (why is this iterator returning Strings instead of str/byte indices?).

But that's separate from this PR and we should merge it first.